File: src/spirite/base/image_data/selection/SelectionEngine.java
package spirite.base.image_data.selection;

import spirite.base.graphics.GraphicsContext;
import spirite.base.graphics.GraphicsContext.Composite;
import spirite.base.graphics.RawImage;
import spirite.base.image_data.ImageWorkspace;
import spirite.base.image_data.UndoEngine;
import spirite.base.image_data.UndoEngine.NullAction;
import spirite.base.image_data.UndoEngine.StackableAction;
import spirite.base.image_data.UndoEngine.UndoableAction;
import spirite.base.image_data.mediums.drawer.IImageDrawer;
import spirite.base.image_data.mediums.drawer.IImageDrawer.IAnchorLiftModule;
import spirite.base.image_data.mediums.drawer.IImageDrawer.ILiftSelectionModule;
import spirite.base.pen.selection_builders.RectSelectionBuilder;
import spirite.base.util.Colors;
import spirite.base.util.ObserverHandler;
import spirite.base.util.linear.MutableTransform;
import spirite.base.util.linear.Rect;
import spirite.base.util.linear.Transform;
import spirite.hybrid.HybridHelper;

import java.util.ArrayList;
import java.util.List;

/***
 *  The SelectionEngine controls the selected image data, moving it from
 *  layer to layer, workspace to workspace, and program-to-clipboard
 *  
 *  A "Selection" is essentially an alpha mask and a corresponding 
 *  BufferedImage which is floating outside of the ImageWorkspace until
 *  it is either merged into existing data or elevated to its own layer.
 *  
 *  I debated on whether or not to integrate offsets into the Selection
 *  rather than have it tracked separately, but offsets would still need
 *  to be tracked for liftedData so there would be no point.
 *  
 * @author Rory Burks
 *
 */
public class SelectionEngine {
	private final ImageWorkspace workspace;
	private final UndoEngine undoEngine;	
	
	
	// Variables relating to Building
	public enum BuildMode {
		DEFAULT, ADD, SUBTRACT, INTERSECTION
	};
	
	// Variables relating to Transforming
	private SelectionMask selection = null;
	private ALiftedData lifted = null;

	
	public SelectionEngine( ImageWorkspace workspace) {
		this.workspace = workspace;
		this.undoEngine = workspace.getUndoEngine();

		workspace.triggerSelectionRefresh();
	}
	
	// ============
	// ==== Basic Gets
	public SelectionMask getSelection() {return selection;}
	public boolean isLifted() {return lifted != null;}
	public ALiftedData getLiftedData() {return lifted;}
	
	// =======
	// ==== Basic and complicated Selection settings
	public void setSelection( final SelectionMask newSelection) {
		final SelectionMask oldSelection = selection;
		
		undoEngine.doAsAggregateAction(() -> {
			finalizeSelection();
			liftedTrans = Transform.Companion.getIdentityMatrix();
			undoEngine.performAndStore( new NullAction() {
				@Override
				protected void undoAction() {
					selection = oldSelection;
					workspace.triggerSelectionRefresh();
				}
				
				@Override
				protected void performAction() {
					selection = newSelection;
					workspace.triggerSelectionRefresh();
				}
			});
		}, "Change Selection");
	}
	public void mergeSelection(SelectionMask newSelection, BuildMode mode) {
		
		switch( mode) {
		case DEFAULT:
			setSelection(newSelection);
			break;
		case ADD:
			setSelection(addSelection(selection, newSelection));
			break;
		case SUBTRACT:
			setSelection(subtractSelection(selection, newSelection));
			break;
		case INTERSECTION:
			setSelection(intersectSelection(selection, newSelection));
			break;
		}
	}
	public SelectionMask addSelection( SelectionMask sel1, SelectionMask sel2) {
		Rect area;
		Rect area1 = (sel1 == null) ? null : new Rect(sel1.ox, sel1.oy, sel1.getWidth(), sel1.getHeight());
		Rect area2 = (sel2 == null) ? null : new Rect(sel2.ox, sel2.oy, sel2.getWidth(), sel2.getHeight());
		
		if( area1 == null && area2 == null)
			return null;
		
		area = (area1 == null)? area2 : area1.union(area2);
		
		RawImage image = HybridHelper.createImage(area.getWidth(), area.getHeight());

		GraphicsContext gc = image.getGraphics();
		gc.preTranslate( -area.getX(), -area.getY());
		if( sel1 != null)
			sel1.drawMask(gc, true);
		if( sel2 != null)
			sel2.drawMask(gc, true);
		
		return new SelectionMask(image, area.getX(), area.getY());
	}
	public SelectionMask subtractSelection( SelectionMask sel1, SelectionMask sel2) {
		if( sel1 == null) return null;
		
		Rect area = new Rect(sel1.ox, sel1.oy, sel1.getWidth(), sel1.getHeight());
		// No need to union with sel2
		
		RawImage image = HybridHelper.createImage(area.getWidth(), area.getHeight());

		GraphicsContext gc = image.getGraphics();
		gc.preTranslate( -area.getX(), -area.getY());
		if( sel1 != null)
			sel1.drawMask(gc, true);
		gc.setComposite( Composite.DST_OUT, 1.0f);
		if( sel2 != null)
			sel2.drawMask(gc, true);

		return new SelectionMask(image, area.getX(), area.getY());
	}
	public SelectionMask intersectSelection( SelectionMask sel1, SelectionMask sel2) {
		if( sel1 == null || sel2 == null) return null;
		
		Rect area = (new Rect(sel1.ox, sel1.oy, sel1.getWidth(), sel1.getHeight()))
				.intersection(new Rect(sel2.ox, sel2.oy, sel2.getWidth(), sel2.getHeight()));
		
		if( area == null || area.isEmpty()) return null;

		RawImage img = HybridHelper.createImage(area.getWidth(), area.getHeight());
		
		GraphicsContext gc = img.getGraphics();
		gc.preTranslate(-area.getX(), -area.getY());
		sel2.drawMask(gc, true);
		gc.setComposite(Composite.DST_IN, 1.0f);
		sel1.drawMask(gc, true);
		//gc.drawImage(img1, 0, 0);
		
		return new SelectionMask(img, area.getX(), area.getY());
	}
	public SelectionMask invertSelection( SelectionMask sel) {
		if( sel == null) return null;
		
		Rect area = new Rect(0,0,workspace.getWidth(), workspace.getHeight());
		
		RawImage image = HybridHelper.createImage( area.getWidth(), area.getHeight());
		GraphicsContext gc = image.getGraphics();
		gc.setColor(Colors.WHITE);
		gc.fillRect(0, 0, area.getWidth(), area.getHeight());
		gc.setComposite(Composite.DST_OUT,  1.0f);
		sel.drawMask(gc, true);
		
		return new SelectionMask(image);
	}
	public SelectionMask buildRectSelection( Rect rect) {
		RectSelectionBuilder rsb = new RectSelectionBuilder(workspace);
		rsb.start(rect.getX(), rect.getY());
		rsb.update(rect.getX()+rect.getWidth(), rect.getY()+rect.getHeight());
		return new SelectionMask( rsb.build());
	}
	
	// =======
	// === Transforming Lifted Data
	Transform liftedTrans = Transform.Companion.getIdentityMatrix();
	Transform proposedLiftedTrans = Transform.Companion.getIdentityMatrix();
	public Transform getLiftedDrawTrans() {return getLiftedDrawTrans(true);}
	public Transform getLiftedDrawTrans( boolean withProposed) {
		MutableTransform trans = new MutableTransform();

		if( isLifted()) {
			trans.preTranslate(-lifted.getWidth()/2, -lifted.getHeight()/2);
			trans.preConcatenate(liftedTrans);
			if( withProposed && proposedLiftedTrans != null)
				trans.preConcatenate(proposedLiftedTrans);
			trans.preTranslate(lifted.getWidth()/2, lifted.getHeight()/2);
		}
		int dx = (selection == null) ? 0 : selection.ox;
		int dy = (selection == null) ? 0 : selection.oy;
		trans.preTranslate(dx, dy);
		
		return trans;
	}
	
	public void transformSelection( Transform trans) {
		liftedTrans = trans.times( liftedTrans);
		workspace.triggerSelectionRefresh();
	}
	
	// ========
	// ==== Proposing Transform
	public void proposeTransform( Transform trans) {
		proposedLiftedTrans = trans;
		workspace.triggerSelectionRefresh();
	}
	public boolean isProposingTransform() {return proposedLiftedTrans != null;}
	public void applyProposedTransform() {
		liftedTrans = proposedLiftedTrans.times(liftedTrans);
		proposedLiftedTrans = null;
		workspace.triggerSelectionRefresh();
	}
	
	// ==========
	// ==== Selection Moving
	public void setOffset(int ox, int oy) {
		List<UndoableAction> actions = undoEngine.aggregateActions(() -> {
			if( !isLifted()) {
				attemptLiftData( workspace.getActiveDrawer());
			}
		});
		
		actions.add(new ModifyingSelection(ox, oy));
		undoEngine.performAndStore(undoEngine.new StackableCompositeAction(actions, "Moving Selection"));
	}

	class ModifyingSelection extends NullAction implements StackableAction 
	{
		int oldOx, oldOy, newOx, newOy;
		
		public ModifyingSelection( int newOx, int newOy) {
			this.newOx = newOx;
			this.newOy = newOy;
			this.oldOx = selection.ox;
			this.oldOy = selection.oy;
		}
		
		@Override protected void performAction() {
			selection = new SelectionMask(selection, newOx, newOy);
			workspace.triggerSelectionRefresh();
		}
		@Override protected void undoAction() {
			selection = new SelectionMask(selection, oldOx, oldOy);
			workspace.triggerSelectionRefresh();
		}
		@Override public boolean canStack(UndoableAction newAction) {
			return newAction instanceof ModifyingSelection;
		}
		@Override public void stackNewAction(UndoableAction newAction) {
			newOx = ((ModifyingSelection)newAction).newOx;
			newOy = ((ModifyingSelection)newAction).newOy;
		}
		@Override
		public String getDescription() {
			return "Moving Selection";
		}
	}
	
	// ============
	// ==== Lifting Data
	public void attemptLiftData( IImageDrawer drawer) {
		
		if( drawer instanceof ILiftSelectionModule) {
			final ALiftedData newLifted = ((ILiftSelectionModule) drawer).liftSelection(selection);
			final ALiftedData oldLifted = lifted;
			liftedTrans = Transform.Companion.getIdentityMatrix();
			
			List<UndoableAction> actions = new ArrayList<>(2);
			
			actions.add(new NullAction() {
				@Override
				protected void performAction() {
					lifted = newLifted;
				}

				@Override
				protected void undoAction() {
					lifted = oldLifted;
				}
			});
			actions.add( new ModifyingSelection(selection.ox, selection.oy));
			
			undoEngine.performAndStore(undoEngine.new StackableCompositeAction(actions, "Lift and Move Data"));
		}
		else
			HybridHelper.beep();
	}
	public void clearLifted() {
		ALiftedData oldLifted = lifted;
		undoEngine.performAndStore(new NullAction() {
			@Override protected void performAction() {
				lifted = null;
				workspace.triggerSelectionRefresh();
			}
			@Override protected void undoAction() {
				lifted = oldLifted;
				workspace.triggerSelectionRefresh();
			}
			@Override public String getDescription() {
				return "Clear Lifted Selection";
			}
		});
	}
	
	// =========
	// ==== Automatic
	private void finalizeSelection() {
		if( isLifted()) {
			undoEngine.doAsAggregateAction(() -> {
				IImageDrawer drawer = workspace.getActiveDrawer();
				if( drawer instanceof IAnchorLiftModule && ((IAnchorLiftModule) drawer).acceptsLifted(lifted)) {
					((IAnchorLiftModule)drawer).anchorLifted(lifted, getLiftedDrawTrans());
				}
				else 
					HybridHelper.beep();
				
				final ALiftedData oldLifted = lifted;
				undoEngine.performAndStore( new NullAction() {
					@Override protected void performAction() {
						lifted = null;
					}
					@Override protected void undoAction() {
						lifted = oldLifted;
					}
				});
			}, "Anchor Lifted Selection");
			
			
			lifted = null;
		}
	}

	public void imageToSelection(RawImage bi, int ox, int oy) {
		undoEngine.doAsAggregateAction(() -> {
			setSelection(buildRectSelection(new Rect(ox, oy, bi.getWidth(), bi.getHeight())));
			
			final FlatLiftedData newLifted = new FlatLiftedData(bi);
			undoEngine.performAndStore(new NullAction() {				
				@Override protected void performAction() {
					lifted = newLifted;
					workspace.triggerSelectionRefresh();
				}
				@Override protected void undoAction() {
					lifted = null;
					workspace.triggerSelectionRefresh();
				}
			});
		}, "Pasted Image to Layer");
	}
	

	// ============
	// ==== Observers
	private final ObserverHandler<MSelectionEngineObserver> selectionObs = new ObserverHandler<>();
    public void addSelectionObserver( MSelectionEngineObserver obs) { selectionObs.addObserver(obs);}
	public void removeSelectionObserver( MSelectionEngineObserver obs) { selectionObs.removeObserver(obs);}
	
	/** SelectionEngineObservers trigger as the selection that's being built
	 * changes and when the Built Selection has changed. */
	public static interface MSelectionEngineObserver {
		public void selectionBuilt(SelectionEvent evt);
		public void buildingSelection( SelectionEvent evt);
	}
	public static class SelectionEvent {
		//Selection selection;	// TODO
	}
	
    void triggerSelectionChanged(SelectionEvent evt) {
    	selectionObs.trigger((MSelectionEngineObserver obs) -> {obs.selectionBuilt(evt);});
    }
    void triggerBuildingSelection(SelectionEvent evt) {
    	selectionObs.trigger((MSelectionEngineObserver obs) -> {obs.buildingSelection(evt);});
    }
}

File: src/main/kotlin/spirite/base/imageData/selection/SelectionEngine.kt
package spirite.base.imageData.selection

import rb.extendo.delegates.DerivedLazy
import rb.glow.img.IImage
import rb.owl.IObservable
import rb.owl.Observable
import rb.vectrix.linear.ITransformF
import rb.vectrix.linear.ImmutableTransformF
import rb.vectrix.mathUtil.d
import rb.vectrix.mathUtil.f
import rbJvm.owl.addWeakObserver
import sgui.core.systems.IImageCreator
import spirite.base.graphics.drawer.IImageDrawer
import spirite.base.graphics.drawer.IImageDrawer.IAnchorLiftModule
import spirite.base.graphics.drawer.IImageDrawer.ILiftSelectionModule
import spirite.base.imageData.IImageWorkspace
import spirite.base.imageData.selection.ISelectionEngine.BuildMode
import spirite.base.imageData.selection.ISelectionEngine.SelectionChangeEvent
import spirite.base.imageData.undo.NullAction
import spirite.base.imageData.undo.StackableAction
import spirite.base.imageData.undo.UndoableAction
import spirite.base.util.linear.Rect
import spirite.base.util.linear.RectangleUtil
import spirite.core.hybrid.DiSet_Hybrid
import spirite.sguiHybrid.Hybrid

interface ISelectionEngine {
    val selectionChangeObserver: IObservable<(SelectionChangeEvent)->Any?>
    val selection : Selection?
    val selectionTransform: ITransformF?
    val selectionExtra: ISelectionExtra?
    fun setSelection(newSelection: Selection?)
    fun mergeSelection(newSelection: Selection, mode: BuildMode)
    fun transformSelection(transform: ITransformF, liftIfEmpty: Boolean = false)
    fun bakeTranslationIntoLifted()

    val liftedData: ILiftedData?
    fun anchorLifted()
    fun clearLifted()
    fun attemptLiftData( drawer: IImageDrawer)
    fun setSelectionWithLifted(newSelection: Selection, lifted: ILiftedData?)
    fun imageToSelection(image: IImage, transform: ITransformF?)

    var proposingTransform: ITransformF?
    fun applyProposingTransform()

    class SelectionChangeEvent(
            val isLiftedChange: Boolean = false
    )

    enum class BuildMode {
        DEFAULT, ADD, SUBTRACT, INTERSECTION
    }
}


class SelectionEngine(
        val workspace: IImageWorkspace,
        private val _imageCreator: IImageCreator = DiSet_Hybrid.imageCreator
) : ISelectionEngine {

    private val selectionDerived = DerivedLazy { selectionMask?.let { Selection(it, selectionTransform) } }
    override val selection by selectionDerived

    private val selectionExtraDerived = DerivedLazy<ISelectionExtra?> {
        val selection = selection ?: return@DerivedLazy null
        (workspace.activeDrawer as? ILiftSelectionModule)?.getSelectionExtra(selection)
    }
    override val selectionExtra: ISelectionExtra? by selectionExtraDerived

    // region Base Selection Stuff
    override var selectionTransform : ITransformF? = null
        private set(value) {
            field = value
            selectionDerived.reset()
        }
    private var selectionMask : IImage? = null
        set(value) {
            field = value
            selectionDerived.reset()
            selectionExtraDerived.reset()
        }

    override fun setSelection(newSelection: Selection?) {
        val newSelection = if( newSelection?.empty == true) null else newSelection
        val liftedData = liftedData
        if( liftedData == null) {
            workspace.undoEngine.performAndStore(ChangeSelectionAction(newSelection))
        }
        else {
            val drawer = workspace.anchorDrawer
            workspace.undoEngine.doAsAggregateAction("Change Selection") {
                val anchorTransform = when(val proposingTransform = proposingTransform) {
                    null -> selectionTransform
                    else -> proposingTransform * (selectionTransform ?: ImmutableTransformF.Identity)
                }
                (drawer as? IAnchorLiftModule)?.apply { if( acceptsLifted(liftedData)) anchorLifted(liftedData, anchorTransform)}

                workspace.undoEngine.performAndStore(object: NullAction() {
                    override val description: String get() = ""
                    override fun performAction() {this@SelectionEngine.liftedData = null}
                    override fun undoAction() {this@SelectionEngine.liftedData = liftedData}
                })

                workspace.undoEngine.performAndStore(ChangeSelectionAction(newSelection))
            }
        }
    }

    inner class ChangeSelectionAction(private val newSelection: Selection?)
        : NullAction()
    {
        private val oldSelection: Selection? = selection
        override val description: String get() = "Change Selection"

        override fun performAction() {
            selectionMask = newSelection?.mask
            selectionTransform = newSelection?.transform
            selectionChangeObserver.trigger { it(SelectionChangeEvent()) }
        }

        override fun undoAction() {
            selectionMask = oldSelection?.mask
            selectionTransform = oldSelection?.transform
            selectionChangeObserver.trigger { it(SelectionChangeEvent()) }
        }
    }

    override fun mergeSelection(newSelection: Selection, mode: BuildMode) {
        if( newSelection.empty) return
        when( mode) {
            BuildMode.DEFAULT -> setSelection(newSelection)
            BuildMode.ADD -> setSelection(selection?.let { it + newSelection} ?: newSelection)
            BuildMode.SUBTRACT -> selection?.let { setSelection(it - newSelection) }
            BuildMode.INTERSECTION -> selection?.let { setSelection(it intersection newSelection) }
        }
    }

    override fun transformSelection(transform: ITransformF, liftIfEmpty: Boolean) {
        val selection = selection ?: return

        val liftedData = liftedData
        val drawer = workspace.activeDrawer
        if( !liftIfEmpty || liftedData != null || drawer !is ILiftSelectionModule)
            workspace.undoEngine.performAndStore(TransformSelectionAction(transform))
        else {
            workspace.undoEngine.doAsAggregateAction("ITransformF", true){
                liftData(drawer, selection)
                workspace.undoEngine.performAndStore(TransformSelectionAction(transform))
            }
        }
    }

    inner class TransformSelectionAction(
            var transform: ITransformF) : NullAction(), StackableAction
    {
        private val originalTransform = selectionTransform

        override val description: String get() = "Moved Selection"

        override fun performAction() {
            selectionTransform = transform * (originalTransform ?: ImmutableTransformF.Identity)
            workspace.activeData?.handle?.refresh()
        }
        override fun undoAction() {
            selectionTransform = originalTransform
            workspace.activeData?.handle?.refresh()
        }

        override fun canStack(other: UndoableAction) = other is TransformSelectionAction
        override fun stackNewAction(other: UndoableAction) {
            transform = (other as TransformSelectionAction).transform * transform
        }

    }

    override fun bakeTranslationIntoLifted() {
        val selection = selection ?: return
        val liftedData = liftedData ?: return
        val selectionMask = selectionMask ?: return
        val transform = selection.transform ?: return

        val baked = liftedData.bake(transform)

        val bakedArea = RectangleUtil.circumscribeTrans(Rect(selectionMask.width, selectionMask.height),transform)
        val newImage = _imageCreator.createImage(bakedArea.width, bakedArea.height)
        val gc = newImage.graphics
        gc.transform = transform
        gc.preTranslate(-bakedArea.x.d, -bakedArea.y.d)
        gc.renderImage(selectionMask, 0.0, 0.0)
        val newSelection = Selection(newImage, ImmutableTransformF.Translation(bakedArea.x.f, bakedArea.y.f) , false)

        workspace.undoEngine.doAsAggregateAction("Bake Lifted") {
            workspace.undoEngine.performAndStore(ChangeSelectionAction(newSelection))
            workspace.undoEngine.performAndStore(ChangeLiftedDataAction(baked))
        }
    }
    // endregion

    override var liftedData: ILiftedData? = null
        private set
    override fun anchorLifted() {
        val liftedData = liftedData ?: return
        val drawer = workspace.anchorDrawer

        if( drawer !is IAnchorLiftModule || drawer.acceptsLifted(liftedData))
            DiSet_Hybrid.beep()
        else
            drawer.anchorLifted(liftedData, selectionTransform)
    }

    override fun setSelectionWithLifted(newSelection: Selection, lifted: ILiftedData?) {
        workspace.undoEngine.doAsAggregateAction("Set Lifted Selection"){
            setSelection(newSelection)
            workspace.undoEngine.performAndStore(ChangeLiftedDataAction(lifted))
        }
    }

    override fun clearLifted() {
        val oldLifted = liftedData
        if( oldLifted != null) {
            workspace.undoEngine.performAndStore( ChangeLiftedDataAction(null))
        }
    }

    override fun attemptLiftData(drawer: IImageDrawer) {
        val selection = selection ?: return
        if( drawer is ILiftSelectionModule)
            liftData(drawer, selection)
        else
            DiSet_Hybrid.beep()
    }

    private fun liftData( drawer: ILiftSelectionModule, selection: Selection) {
        workspace.undoEngine.doAsAggregateAction("Lifted Data") {
            workspace.undoEngine.performAndStore(ChangeLiftedDataAction(drawer.liftSelection(selection)))
        }
    }

    inner class ChangeLiftedDataAction(val new: ILiftedData?) : NullAction() {
        val old = liftedData
        override val description: String get() = "Changed Lifted Data"

        override fun performAction() {
            liftedData = new
            selectionChangeObserver.trigger { it(SelectionChangeEvent(true)) }
        }

        override fun undoAction() {
            liftedData = old
            selectionChangeObserver.trigger { it(SelectionChangeEvent(true)) }
        }
    }

    override fun imageToSelection(image: IImage, transform: ITransformF?) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override var proposingTransform: ITransformF? = null
        set(value) {
            field = value
            selectionDerived.reset()
            selectionChangeObserver.trigger { it(SelectionChangeEvent(true)) }
        }

    override fun applyProposingTransform() {
        // Note: As of now it will bake any exisiting selection transform in addition to the proposing transform
        selectionMask ?: return
        val proposingTransform = proposingTransform ?: return
        val selectionTransform = selectionTransform
        val newTransform = if( selectionTransform == null) proposingTransform else proposingTransform * selectionTransform
        this.selectionTransform = newTransform
        bakeTranslationIntoLifted()
    }

    override val selectionChangeObserver = Observable<(SelectionChangeEvent)->Any?>()

    // For Selection Extra trigger-on-change
    private val _workspaceDrawerK = workspace.activeDrawerObs.addWeakObserver { selectionExtraDerived.reset() }
}
